File: rust/src/medium/knapsack.rs
pub struct Item {
    pub weight: u32,
    pub value: u32,
}

pub fn maximum_value(_max_weight: u32, _items: Vec<Item>) -> u32 {
    unimplemented!("Solve the knapsack exercise");
}

#[ignore]
#[test]
fn test_example_knapsack() {
    let max_weight = 10;

    let items = vec![
        Item {
            weight: 5,

            value: 10,
        },
        Item {
            weight: 4,

            value: 40,
        },
        Item {
            weight: 6,

            value: 30,
        },
        Item {
            weight: 4,

            value: 50,
        },
    ];

    assert_eq!(maximum_value(max_weight, items), 90);
}

File: elixir/lib/medium/knapsack.ex
defmodule Knapsack do
  def maximum_value([], _max_weight), do: 0

  def maximum_value(items, max_weight) do
    cache =
      Enum.reduce(1..length(items), %{}, fn item_index, cache ->
        iterate(item_index, items, max_weight, cache)
      end)

    cache[length(items)][max_weight] || 0
  end

  defp iterate(item_index, items, max_weight, cache) do
    item = Enum.at(items, item_index - 1)

    Enum.reduce(0..max_weight, cache, fn weight, cache ->
      value_without_item = cache[item_index - 1][weight] || 0

      value_with_item =
        if weight >= item.weight do
          remainder_value = cache[item_index - 1][weight - item.weight] || 0

          item.value + remainder_value
        else
          0
        end

      max_value = max(value_without_item, value_with_item)

      put_in(cache, [Access.key(item_index, %{}), weight], max_value)
    end)
  end
end
